runtime/src/accounts_db.rs

```diff
@@ -879,6 +881,150 @@ impl AccountsDB {
+            stored_accounts
+                .iter()
+                .filter(
+                    |(pubkey, _account, _storage_size, (store_id, offset), _write_version)| {
+                        if let Some((list, _)) = accounts_index.get(pubkey, &no_ancestors) {
+                            list.iter()
+                                .any(|(_slot, i)| i.store_id == *store_id && i.offset == *offset)
+                        } else {
+                            false
+                        }
+                    },
+                )
+                .collect()
+        };
+
+        let alive_total: u64 = alive_accounts
+            .iter()
+            .map(|(_pubkey, _account, account_size, _location, _write_verion)| *account_size as u64)
+            .sum();
+        let aligned_total: u64 = (alive_total + (PAGE_SIZE - 1)) & !(PAGE_SIZE - 1);
+
+        debug!(
+            "shrinking: slot: {}, stored_accounts: {} => alive_accounts: {} ({} bytes; aligned to: {})",
+            slot,
+            stored_accounts.len(),
+            alive_accounts.len(),
+            alive_total,
+            aligned_total
+        );
+
+        if aligned_total > 0 {
+            let mut accounts = Vec::with_capacity(alive_accounts.len());
+            let mut hashes = Vec::with_capacity(alive_accounts.len());
+            let mut write_versions = Vec::with_capacity(alive_accounts.len());
+
+            for (pubkey, account, _size, _location, write_version) in alive_accounts {
+                accounts.push((pubkey, account));
+                hashes.push(account.hash);
+                write_versions.push(*write_version);
+            }
+
+            let shrunken_store = self.create_and_insert_store(slot, aligned_total);
+
+            // here, we're writing back alive_accounts. That should be an atomic operation
+            // without use of rather wide locks in this whole function, because we're
+            // mutating rooted slots; There should be no writers to them.
+            let infos = self.store_accounts_to(
+                slot,
+                &accounts,
+                &hashes,
+                |_| shrunken_store.clone(),
+                write_versions.into_iter(),
+            );
+            let reclaims = self.update_index(slot, infos, &accounts);
+
+            self.handle_reclaims(&reclaims);
+
+            let mut storage = self.storage.write().unwrap();
+            if let Some(slot_storage) = storage.0.get_mut(&slot) {
+                slot_storage.retain(|_key, store| store.count() > 0);
+            }
+        }
+    }
+
+    // Infinitely returns rooted roots in cyclic order
+    fn next_shrink_slot(&self) -> Option<Slot> {
+        let next = {
+            let mut candidates = self.shrink_candidate_slots.lock().unwrap();
+            candidates.pop()
+        };
+
+        if next.is_some() {
+            next
+        } else {
+            let mut new_all_slots = self.all_root_slots_in_index();
```

Comment:
After local testing, I've found this logic is too naive for real hostile environment. But first I'll merge this...